Repeated release builds that use Android flavors and flutter_localizations fail with gradle error

### Steps to reproduce

1. Create new Flutter project (I used latest Flutter 3.29)
2. Add flavors to /android/app/build.gradle.kts
```kotlin
   flavorDimensions += "flavor-type" 
    productFlavors {
        create("dev") {
            dimension = "flavor-type"
            applicationId = "com.Li3NK.spark.dev"
            resValue("string", "app_name", "Spark dev")
        }
        create("prod") {
            dimension = "flavor-type"
            applicationId = "com.Li3NK.spark"
            resValue("string", "app_name", "Spark")
        }
    }
```
3. Add l10n.yaml file to the root of the project
`arb-dir: lib/l10n
template-arb-file: app_en.arb
output-localization-file: app_localizations.dart
synthetic-package: false`
4. Add /lib/l10n/app_en.arb file
`{
  "commonYes": "Yes"
}`
5. Generate the localization files with flutter gen-l10n if they are not generated automatically
6. Modify main.dart to use the localized string
```dart
import 'package:flutter/material.dart';
import 'package:test_329_create/l10n/app_localizations.dart';

void main() {
  runApp(const MainApp());
}

class MainApp extends StatelessWidget {
  const MainApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(body: Center(child: Text(AppLocalizations.of(context)!.commonYes))),
    );
  }
}
```
7. Run flutter build apk --release --flavor=dev once. It should pass succesfully
8. Run the same command again

### Expected results

Second build is also successful

### Actual results

Second build fails with gradle exception (see logs below)

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:bug/l10n/app_localizations.dart';

void main() {
  runApp(const MainApp());
}

class MainApp extends StatelessWidget {
  const MainApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(body: Center(child: Text(AppLocalizations.of(context)!.commonYes))),
    );
  }
}

```

</details>


### Screenshots or Video

_No response_

### Logs

<details open><summary>Logs</summary>

```console
`
Running Gradle task 'assembleDevRelease'...                          
FAILURE: Build failed with an exception.

* What went wrong:
Some problems were found with the configuration of task ':app:compileFlutterBuildProdRelease' (type 'FlutterTask').
  - Gradle detected a problem with the following location: 'D:\Misc\Code\test_329_create\lib\l10n\app_localizations_en.dart'.

    Reason: Task ':app:compileFlutterBuildDevRelease' uses this output of task ':app:compileFlutterBuildProdRelease' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.       

    Possible solutions:
      1. Declare task ':app:compileFlutterBuildProdRelease' as an input of ':app:compileFlutterBuildDevRelease'.
      2. Declare an explicit dependency on ':app:compileFlutterBuildProdRelease' from ':app:compileFlutterBuildDevRelease' using Task#dependsOn.
      3. Declare an explicit dependency on ':app:compileFlutterBuildProdRelease' from ':app:compileFlutterBuildDevRelease' using Task#mustRunAfter.
`
```

</details>


### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
└ $ flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, 3.29.0, on Microsoft Windows [Version
    10.0.26100.3194], locale uk-UA)
[√] Windows Version (11 Pro 64-bit, 24H2, 2009)
[√] Android toolchain - develop for Android devices (Android SDK
    version 35.0.0)
[√] Chrome - develop for the web
[√] Visual Studio - develop Windows apps (Visual Studio Community
    2022 17.12.3)
[√] Android Studio (version 2024.2)
[√] VS Code (version 1.97.2)
[√] Connected device (3 available)
[√] Network resources

• No issues found!
                                                      pwsh | 10:36:10
┌ S9Th@S9Th-PC  test_329_create 
└ $ flutter doctor -v 
[√] Flutter (Channel stable, 3.29.0, on Microsoft Windows [Version
    10.0.26100.3194], locale uk-UA) [367ms]
    • Flutter version 3.29.0 on channel stable at
      D:\Misc\Code\env\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 35c388afb5 (9 days ago), 2025-02-10 12:48:41
      -0800
    • Engine revision f73bfc4522
    • Dart version 3.7.0
    • DevTools version 2.42.2

[√] Windows Version (11 Pro 64-bit, 24H2, 2009) [4,6s]

[√] Android toolchain - develop for Android devices (Android SDK
    version 35.0.0) [1 850ms]
    • Android SDK at D:\Misc\Code\env\AndroidSDK
    • Platform android-35, build-tools 35.0.0
    • ANDROID_HOME = D:\Misc\Code\env\AndroidSDK
    • Java binary at: C:\Program Files\Android\Android
      Studio\jbr\bin\java
      This is the JDK bundled with the latest Android Studio
      installation on this machine.
      To manually set the JDK path, use: `flutter config
      --jdk-dir="path/to/jdk"`.
    • Java version OpenJDK Runtime Environment (build
      21.0.4+-12508038-b607.1)
    • All Android licenses accepted.

[√] Chrome - develop for the web [84ms]
    • Chrome at C:\Program Files
      (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop Windows apps (Visual Studio Community
    2022 17.12.3) [83ms]
    • Visual Studio at C:\Program Files\Microsoft Visual
      Studio\2022\Community
    • Visual Studio Community 2022 version 17.12.35527.113
    • Windows 10 SDK version 10.0.19041.0

[√] Android Studio (version 2024.2) [21ms]
    • Android Studio at C:\Program Files\Android\Android Studio      
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      21.0.4+-12508038-b607.1)

[√] VS Code (version 1.97.2) [19ms]
    • VS Code at C:\Users\S9Th\AppData\Local\Programs\Microsoft VS
      Code
    • Flutter extension version 3.104.0

[√] Connected device (3 available) [224ms]
    • Windows (desktop) • windows • windows-x64    • Microsoft       
      Windows [Version 10.0.26100.3194]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome   
      132.0.6834.197
    • Edge (web)        • edge    • web-javascript • Microsoft Edge  
      133.0.3065.69

[√] Network resources [342ms]
    • All expected network resources are available.

• No issues found!
```

</details>

@matanlurey just in case it's still needed. Yes I can reproduce it on your repo as well
```console
┌ S9Th@S9Th-PC  163591-flavor-town   main ≡  error 
└ $ flutter build apk --flavor=dev

Running Gradle task 'assembleDevRelease'...                       150,0s
√ Built build\app\outputs\flutter-apk\app-dev-release.apk (16.6MB)
                                                                         pwsh | 09:13:31
┌ S9Th@S9Th-PC  163591-flavor-town   main ≡ 
└ $ flutter build apk --flavor=dev


FAILURE: Build failed with an exception.

* What went wrong:
Some problems were found with the configuration of task ':app:compileFlutterBuildProdRelease' (type 'FlutterTask').
  - Gradle detected a problem with the following location: 'D:\Misc\Code\163591-flavor-town\lib\l10n\app_localizations_en.dart'.

    Reason: Task ':app:compileFlutterBuildDevRelease' uses this output of task ':app:compileFlutterBuildProdRelease' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
```

My first build after a flutter clean always works (using Windows 11), this problem only occurs when I try to build an apk or appbundle the second time.

<details>
   <summary>Terminal Logs</summary>

   ```PowerShell
flutter build appbundle --release --no-pub --suppress-analytics --build-number=4 --build-name=1.44.0 --flavor=staging --target=lib/main.dart --obfuscate --split-debug-info=./build/app/outputs/symbols

Font asset "MaterialIcons-Regular.otf" was tree-shaken, reducing it from 1645184 to 11532 bytes (99.3% reduction). Tree-shaking can be disabled by providing the --no-tree-shake-icons flag when building your app.
Font asset "CupertinoIcons.ttf" was tree-shaken, reducing it from 257628 to 848 bytes (99.7% reduction). Tree-shaking can be disabled by providing the --no-tree-shake-icons flag when building your app.

FAILURE: Build failed with an exception.

* What went wrong:
Some problems were found with the configuration of task ':app:compileFlutterBuildProductionRelease' (type 'FlutterTask').
  - Gradle detected a problem with the following location: 'C:\Users\sample\projects\project_name\lib\l10n\generated\app_localizations.dart'.       

    Reason: Task ':app:compileFlutterBuildStagingRelease' uses this output of task ':app:compileFlutterBuildProductionRelease' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.

    Possible solutions:
      1. Declare task ':app:compileFlutterBuildProductionRelease' as an input of ':app:compileFlutterBuildStagingRelease'.
      2. Declare an explicit dependency on ':app:compileFlutterBuildProductionRelease' from ':app:compileFlutterBuildStagingRelease' using Task#dependsOn.
      3. Declare an explicit dependency on ':app:compileFlutterBuildProductionRelease' from ':app:compileFlutterBuildStagingRelease' using Task#mustRunAfter.

    For more information, please refer to https://docs.gradle.org/8.10.2/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.
  - Gradle detected a problem with the following location: 'C:\Users\sample\projects\project_name\lib\l10n\generated\app_localizations_en.dart'.    

    Reason: Task ':app:compileFlutterBuildStagingRelease' uses this output of task ':app:compileFlutterBuildProductionRelease' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.

    Possible solutions:
      1. Declare task ':app:compileFlutterBuildProductionRelease' as an input of ':app:compileFlutterBuildStagingRelease'.
      2. Declare an explicit dependency on ':app:compileFlutterBuildProductionRelease' from ':app:compileFlutterBuildStagingRelease' using Task#dependsOn.
      3. Declare an explicit dependency on ':app:compileFlutterBuildProductionRelease' from ':app:compileFlutterBuildStagingRelease' using Task#mustRunAfter.

    For more information, please refer to https://docs.gradle.org/8.10.2/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.
  - Gradle detected a problem with the following location: 'C:\Users\sample\projects\project_name\lib\l10n\generated\app_localizations.dart'.       

    Reason: Task ':app:compileFlutterBuildProductionRelease' uses this output of task ':app:compileFlutterBuildStagingRelease' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.

    Possible solutions:
      1. Declare task ':app:compileFlutterBuildStagingRelease' as an input of ':app:compileFlutterBuildProductionRelease'.
      2. Declare an explicit dependency on ':app:compileFlutterBuildStagingRelease' from ':app:compileFlutterBuildProductionRelease' using Task#dependsOn.
      3. Declare an explicit dependency on ':app:compileFlutterBuildStagingRelease' from ':app:compileFlutterBuildProductionRelease' using Task#mustRunAfter.

    For more information, please refer to https://docs.gradle.org/8.10.2/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.
  - Gradle detected a problem with the following location: 'C:\Users\sample\projects\project_name\lib\l10n\generated\app_localizations_en.dart'.

    Reason: Task ':app:compileFlutterBuildProductionRelease' uses this output of task ':app:compileFlutterBuildStagingRelease' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.

    Possible solutions:
      1. Declare task ':app:compileFlutterBuildStagingRelease' as an input of ':app:compileFlutterBuildProductionRelease'.
      2. Declare an explicit dependency on ':app:compileFlutterBuildStagingRelease' from ':app:compileFlutterBuildProductionRelease' using Task#dependsOn.
      3. Declare an explicit dependency on ':app:compileFlutterBuildStagingRelease' from ':app:compileFlutterBuildProductionRelease' using Task#mustRunAfter.

    For more information, please refer to https://docs.gradle.org/8.10.2/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

BUILD FAILED in 28s
Running Gradle task 'bundleStagingRelease'...                      29.3s
Gradle task bundleStagingRelease failed with exit code 1
   ```

</details>

--- a/packages/flutter_tools/lib/src/build_system/targets/localizations.dart
+++ b/packages/flutter_tools/lib/src/build_system/targets/localizations.dart
@@ -58,6 +58,7 @@ class GenerateLocalizationsTarget extends Target {
     final LocalizationOptions options = parseLocalizationsOptionsFromYAML(
       file: configFile,
       logger: environment.logger,
+      fileSystem: environment.fileSystem,
       defaultArbDir: defaultArbDir,
       defaultSyntheticPackage: !featureFlags.isExplicitPackageDependenciesEnabled,
     );

--- a/packages/flutter_tools/lib/src/commands/generate_localizations.dart
+++ b/packages/flutter_tools/lib/src/commands/generate_localizations.dart
@@ -260,6 +260,7 @@ class GenerateLocalizationsCommand extends FlutterCommand {
       options = parseLocalizationsOptionsFromYAML(
         file: _fileSystem.file('l10n.yaml'),
         logger: _logger,
+        fileSystem: _fileSystem,
         defaultArbDir: defaultArbDir,
         defaultSyntheticPackage: !featureFlags.isExplicitPackageDependenciesEnabled,
       );

--- a/packages/flutter_tools/lib/src/localizations/gen_l10n.dart
+++ b/packages/flutter_tools/lib/src/localizations/gen_l10n.dart
@@ -1000,6 +1000,14 @@ class LocalizationsGenerator {
     return true;
   }
 
+  void _addToFileList(List<String> fileList, String path) {
+    fileList.add(_fs.path.normalize(path));
+  }
+
+  void _addAllToFileList(List<String> fileList, Iterable<String> paths) {
+    fileList.addAll(paths.map(_fs.path.normalize));
+  }
+
   // Load _allMessages from templateArbFile and _allBundles from all of the ARB
   // files in inputDirectory. Also initialized: supportedLocales.
   void loadResources() {
@@ -1030,7 +1038,8 @@ class LocalizationsGenerator {
             .toList();
     hadErrors = _allMessages.any((Message message) => message.hadErrors);
     if (inputsAndOutputsListFile != null) {
-      _inputFileList.addAll(
+      _addAllToFileList(
+        _inputFileList,
         _allBundles.bundles.map((AppResourceBundle bundle) {
           return bundle.file.absolute.path;
         }),
@@ -1492,7 +1501,7 @@ The plural cases must be one of "=0", "=1", "=2", "zero", "one", "two", "few", "
     // Generate the required files for localizations.
     _languageFileMap.forEach((File file, String contents) {
       file.writeAsStringSync(useCRLF ? contents.replaceAll('\n', '\r\n') : contents);
-      _outputFileList.add(file.absolute.path);
+      _addToFileList(_outputFileList, file.absolute.path);
     });
 
     baseOutputFile.writeAsStringSync(
@@ -1527,7 +1536,7 @@ The plural cases must be one of "=0", "=1", "=2", "zero", "one", "two", "few", "
       );
     }
     final File? inputsAndOutputsListFileLocal = inputsAndOutputsListFile;
-    _outputFileList.add(baseOutputFile.absolute.path);
+    _addToFileList(_outputFileList, baseOutputFile.absolute.path);
     if (inputsAndOutputsListFileLocal != null) {
       // Generate a JSON file containing the inputs and outputs of the gen_l10n script.
       if (!inputsAndOutputsListFileLocal.existsSync()) {
@@ -1548,7 +1557,7 @@ The plural cases must be one of "=0", "=1", "=2", "zero", "one", "two", "few", "
   void _generateUntranslatedMessagesFile(Logger logger, File untranslatedMessagesFile) {
     if (_unimplementedMessages.isEmpty) {
       untranslatedMessagesFile.writeAsStringSync('{}');
-      _outputFileList.add(untranslatedMessagesFile.absolute.path);
+      _addToFileList(_outputFileList, untranslatedMessagesFile.absolute.path);
       return;
     }
 
@@ -1576,6 +1585,6 @@ The plural cases must be one of "=0", "=1", "=2", "zero", "one", "two", "few", "
 
     resultingFile += '}\n';
     untranslatedMessagesFile.writeAsStringSync(resultingFile);
-    _outputFileList.add(untranslatedMessagesFile.absolute.path);
+    _addToFileList(_outputFileList, untranslatedMessagesFile.absolute.path);
   }
 }

--- a/packages/flutter_tools/lib/src/localizations/localizations_utils.dart
+++ b/packages/flutter_tools/lib/src/localizations/localizations_utils.dart
@@ -479,6 +479,7 @@ class LocalizationOptions {
 LocalizationOptions parseLocalizationsOptionsFromYAML({
   required File file,
   required Logger logger,
+  required FileSystem fileSystem,
   required String defaultArbDir,
   required bool defaultSyntheticPackage,
 }) {
@@ -497,14 +498,24 @@ LocalizationOptions parseLocalizationsOptionsFromYAML({
     throw Exception();
   }
   return LocalizationOptions(
-    arbDir: _tryReadUri(yamlNode, 'arb-dir', logger)?.path ?? defaultArbDir,
-    outputDir: _tryReadUri(yamlNode, 'output-dir', logger)?.path,
-    templateArbFile: _tryReadUri(yamlNode, 'template-arb-file', logger)?.path,
-    outputLocalizationFile: _tryReadUri(yamlNode, 'output-localization-file', logger)?.path,
-    untranslatedMessagesFile: _tryReadUri(yamlNode, 'untranslated-messages-file', logger)?.path,
+    arbDir: _tryReadFilePath(yamlNode, 'arb-dir', logger, fileSystem) ?? defaultArbDir,
+    outputDir: _tryReadFilePath(yamlNode, 'output-dir', logger, fileSystem),
+    templateArbFile: _tryReadFilePath(yamlNode, 'template-arb-file', logger, fileSystem),
+    outputLocalizationFile: _tryReadFilePath(
+      yamlNode,
+      'output-localization-file',
+      logger,
+      fileSystem,
+    ),
+    untranslatedMessagesFile: _tryReadFilePath(
+      yamlNode,
+      'untranslated-messages-file',
+      logger,
+      fileSystem,
+    ),
     outputClass: _tryReadString(yamlNode, 'output-class', logger),
     header: _tryReadString(yamlNode, 'header', logger),
-    headerFile: _tryReadUri(yamlNode, 'header-file', logger)?.path,
+    headerFile: _tryReadFilePath(yamlNode, 'header-file', logger, fileSystem),
     useDeferredLoading: _tryReadBool(yamlNode, 'use-deferred-loading', logger),
     preferredSupportedLocales: _tryReadStringList(yamlNode, 'preferred-supported-locales', logger),
     syntheticPackage:
@@ -596,8 +607,8 @@ List<String>? _tryReadStringList(YamlMap yamlMap, String key, Logger logger) {
   throw Exception();
 }
 
-// Try to read a valid `Uri` or null from `yamlMap`, otherwise throw.
-Uri? _tryReadUri(YamlMap yamlMap, String key, Logger logger) {
+// Try to read a valid file `Uri` or null from `yamlMap` to file path, otherwise throw.
+String? _tryReadFilePath(YamlMap yamlMap, String key, Logger logger, FileSystem fileSystem) {
   final String? value = _tryReadString(yamlMap, key, logger);
   if (value == null) {
     return null;
@@ -606,5 +617,5 @@ Uri? _tryReadUri(YamlMap yamlMap, String key, Logger logger) {
   if (uri == null) {
     logger.printError('"$value" must be a relative file URI');
   }
-  return uri;
+  return uri != null ? fileSystem.path.normalize(uri.path) : null;
 }

--- a/packages/flutter_tools/test/commands.shard/hermetic/generate_localizations_test.dart
+++ b/packages/flutter_tools/test/commands.shard/hermetic/generate_localizations_test.dart
@@ -7,6 +7,7 @@ import 'package:flutter_tools/src/artifacts.dart';
 import 'package:flutter_tools/src/base/file_system.dart';
 import 'package:flutter_tools/src/base/logger.dart';
 import 'package:flutter_tools/src/build_system/build_system.dart';
+import 'package:flutter_tools/src/build_system/depfile.dart' show Depfile;
 import 'package:flutter_tools/src/build_system/targets/localizations.dart';
 import 'package:flutter_tools/src/cache.dart';
 import 'package:flutter_tools/src/commands/generate_localizations.dart';
@@ -515,6 +516,42 @@ format: true
       expect(processManager, hasNoRemainingExpectations);
     },
   );
+
+  testUsingContext('generates normalized input & output file paths', () async {
+    final File arbFile = fileSystem.file(fileSystem.path.join('lib', 'l10n', 'app_en.arb'))
+      ..createSync(recursive: true);
+    arbFile.writeAsStringSync('''
+{
+  "helloWorld": "Hello, World!"
+}''');
+    final File configFile = fileSystem.file('l10n.yaml')..createSync();
+    // Writing both forward and backward slashes to test both cases.
+    configFile.writeAsStringSync(r'''
+arb-dir: lib/l10n
+output-dir: lib\l10n
+format: false
+''');
+    final File pubspecFile = fileSystem.file('pubspec.yaml')..createSync();
+    pubspecFile.writeAsStringSync(BasicProjectWithFlutterGen().pubspec);
+
+    processManager.addCommand(const FakeCommand(command: <String>[]));
+    final Environment environment = Environment.test(
+      fileSystem.currentDirectory,
+      artifacts: artifacts,
+      processManager: processManager,
+      fileSystem: fileSystem,
+      logger: BufferLogger.test(),
+    );
+    const Target buildTarget = GenerateLocalizationsTarget();
+    await buildTarget.build(environment);
+
+    final File dependencyFile = environment.buildDir.childFile(buildTarget.depfiles.single);
+    final Depfile depfile = environment.depFileService.parse(dependencyFile);
+
+    final String oppositeSeparator = fileSystem.path.separator == '/' ? r'\' : '/';
+    expect(depfile.inputs, everyElement(isNot(contains(oppositeSeparator))));
+    expect(depfile.outputs, everyElement(isNot(contains(oppositeSeparator))));
+  });
 
   testUsingContext(
     'nullable-getter defaults to true',

--- a/packages/flutter_tools/test/general.shard/build_system/targets/localizations_test.dart
+++ b/packages/flutter_tools/test/general.shard/build_system/targets/localizations_test.dart
@@ -51,6 +51,7 @@ nullable-getter: false
     final LocalizationOptions options = parseLocalizationsOptionsFromYAML(
       file: configFile,
       logger: BufferLogger.test(),
+      fileSystem: fileSystem,
       defaultArbDir: fileSystem.path.join('lib', 'l10n'),
       defaultSyntheticPackage: true,
     );
@@ -90,6 +91,7 @@ nullable-getter: false
     final LocalizationOptions options = parseLocalizationsOptionsFromYAML(
       file: configFile,
       logger: BufferLogger.test(),
+      fileSystem: fileSystem,
       defaultArbDir: fileSystem.path.join('lib', 'l10n'),
       defaultSyntheticPackage: true,
     );
@@ -118,6 +120,7 @@ nullable-getter: false
     final LocalizationOptions options = parseLocalizationsOptionsFromYAML(
       file: configFile,
       logger: BufferLogger.test(),
+      fileSystem: fileSystem,
       defaultArbDir: fileSystem.path.join('lib', 'l10n'),
       defaultSyntheticPackage: false,
     );
@@ -136,6 +139,7 @@ preferred-supported-locales: ['en_US', 'de']
       final LocalizationOptions options = parseLocalizationsOptionsFromYAML(
         file: configFile,
         logger: BufferLogger.test(),
+        fileSystem: fileSystem,
         defaultArbDir: fileSystem.path.join('lib', 'l10n'),
         defaultSyntheticPackage: true,
       );
@@ -156,6 +160,7 @@ use-deferred-loading: string
         () => parseLocalizationsOptionsFromYAML(
           file: configFile,
           logger: BufferLogger.test(),
+          fileSystem: fileSystem,
           defaultArbDir: fileSystem.path.join('lib', 'l10n'),
           defaultSyntheticPackage: true,
         ),
@@ -174,6 +179,7 @@ template-arb-file: {name}_en.arb
       () => parseLocalizationsOptionsFromYAML(
         file: configFile,
         logger: BufferLogger.test(),
+        fileSystem: fileSystem,
         defaultArbDir: fileSystem.path.join('lib', 'l10n'),
         defaultSyntheticPackage: true,
       ),

--- a/packages/flutter_tools/test/general.shard/generate_localizations_test.dart
+++ b/packages/flutter_tools/test/general.shard/generate_localizations_test.dart
@@ -7,6 +7,7 @@ import 'package:flutter_tools/src/artifacts.dart';
 import 'package:flutter_tools/src/base/file_system.dart';
 import 'package:flutter_tools/src/base/logger.dart';
 import 'package:flutter_tools/src/convert.dart';
+import 'package:flutter_tools/src/globals.dart' as globals show platform;
 import 'package:flutter_tools/src/localizations/gen_l10n.dart';
 import 'package:flutter_tools/src/localizations/gen_l10n_types.dart';
 import 'package:flutter_tools/src/localizations/localizations_utils.dart';
@@ -89,8 +90,10 @@ void main() {
     bool relaxSyntax = false,
     bool useNamedParameters = false,
     void Function(Directory)? setup,
+    FileSystem? fileSystem,
   }) {
-    final Directory l10nDirectory = fs.directory(defaultL10nPath)..createSync(recursive: true);
+    final Directory l10nDirectory = (fileSystem ?? fs).directory(defaultL10nPath)
+      ..createSync(recursive: true);
     for (final String locale in localeToArbFile.keys) {
       l10nDirectory.childFile('app_$locale.arb').writeAsStringSync(localeToArbFile[locale]!);
     }
@@ -98,7 +101,7 @@ void main() {
       setup(l10nDirectory);
     }
     return LocalizationsGenerator(
-        fileSystem: fs,
+        fileSystem: fileSystem ?? fs,
         inputPathString: l10nDirectory.path,
         outputPathString: outputPathString ?? l10nDirectory.path,
         templateArbFileName: defaultTemplateArbFileName,
@@ -611,6 +614,29 @@ void main() {
       outputList,
       contains(fs.path.absolute(defaultL10nPath, 'output-localization-file_es.dart')),
     );
+  });
+
+  testUsingContext('generates normalized input & output file paths', () {
+    final FileSystem fs = MemoryFileSystem.test(
+      style: globals.platform.isWindows ? FileSystemStyle.windows : FileSystemStyle.posix,
+    );
+    setupLocalizations(
+      <String, String>{'en': singleMessageArbFileString, 'es': singleEsMessageArbFileString},
+      fileSystem: fs,
+      inputsAndOutputsListPath: defaultL10nPath,
+    );
+    final File inputsAndOutputsList = fs.file(
+      fs.path.join(defaultL10nPath, 'gen_l10n_inputs_and_outputs.json'),
+    );
+    expect(inputsAndOutputsList.existsSync(), isTrue);
+
+    final Map<String, dynamic> jsonResult =
+        json.decode(inputsAndOutputsList.readAsStringSync()) as Map<String, dynamic>;
+    final String oppositeSeparator = globals.platform.isWindows ? '/' : r'\';
+    final List<dynamic> inputList = jsonResult['inputs'] as List<dynamic>;
+    expect(inputList, everyElement(isNot(contains(oppositeSeparator))));
+    final List<dynamic> outputList = jsonResult['outputs'] as List<dynamic>;
+    expect(outputList, everyElement(isNot(contains(oppositeSeparator))));
   });
 
   testWithoutContext('setting both a headerString and a headerFile should fail', () {

